### PR TITLE
chains: add map reduce documents and apply

### PR DESCRIPTION
**Pull request by FluffyKebab**
This adds the map reduce document combing chain and the apply function to run a chain with multiple inputs at the same time. 
https://github.com/hwchase17/langchainjs/blob/main/langchain/src/chains/combine_docs_chain.ts
https://github.com/hwchase17/langchainjs/blob/main/langchain/src/chains/base.ts

### PR Checklist

- [x] Read the [Contributing documentation](https://github.com/tmc/langchaingo/blob/main/CONTRIBUTING.md).
- [x] Read the [Code of conduct documentation](https://github.com/tmc/langchaingo/blob/main/CODE_OF_CONDUCT.md).
- [x] Name your Pull Request title clearly, concisely, and prefixed with the name of the primarily affected package you changed according to [Good commit messages](https://go.dev/doc/contribute#commit_messages) (such as `memory: add interfaces for X, Y` or `util: add whizzbang helpers`).
- [x] Check that there isn't already a PR that solves the problem the same way to avoid creating a duplicate.
- [x] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue that it solves (e.g. `Fixes #123`).
- [x] Describes the source of new concepts.
- [x] References existing implementations as appropriate.
- [x] Contains test coverage for new functions.
- [x] Passes all [`golangci-lint`](https://golangci-lint.run/) checks.

### CODE REVIEW

File: chains/chains.go [hunk 1]
```diff
@@ -121,6 +121,53 @@ func Predict(ctx context.Context, c Chain, inputValues map[string]any, options .
...
+
+	inputJobs := make(chan map[string]any, len(inputValues))
+	resultsChan := make(chan struct {
+		result map[string]any
+		err    error
+	}, len(inputValues))
+	defer close(inputJobs)
+	defer close(resultsChan)
+
+	for w := 0; w < maxWorkers; w++ {
```
Comment (tmc) [inline chains/chains.go:140]: can you double check this doesn't suffer from the go gotcha https://github.com/golang/go/issues/60078

**Review by tmc — APPROVED** [review]
great work!

File: chains/chains.go (hunk 1 above)
Comment (FluffyKebab) [reply chains/chains.go:140]: I dont think it does. The variable w is not used in the anonymous function.

File: chains/chains.go [hunk 2]
```diff
@@ -121,6 +121,53 @@ func Predict(ctx context.Context, c Chain, inputValues map[string]any, options .
...
+	}
+
+	results := make([]map[string]any, len(inputValues))
+	for i := range inputValues {
+		r := <-resultsChan
+		if r.err != nil {
+			return nil, r.err
+		}
+		results[i] = r.result
+	}
```
Comment (tobiade) [inline chains/chains.go:166]: How do we know what output corresponds to a particular input? This seems to be storing results in random order. I think we should be passing the index of the input when we write to `inputJobs`.

In `mapResultsToLLMChainInputs` I think you could possibly be mapping wrong results to inputs.

I'm also assuming this `Apply` function can be used for things other than MapReduce, so we'd need to know the output for a particular input.

I could be missing something though - thoughts?

Comment (FluffyKebab) [reply chains/chains.go:166]: Yes you are correct. I will make an update for this.

File: chains/chains.go [hunk 3]
```diff
@@ -121,6 +121,53 @@ func Predict(ctx context.Context, c Chain, inputValues map[string]any, options .
...
+	if maxWorkers <= 0 {
+		maxWorkers = _defaultApplyMaxNumberWorkers
+	}
+
+	inputJobs := make(chan map[string]any, len(inputValues))
+	resultsChan := make(chan struct {
+		result map[string]any
+		err    error
+	}, len(inputValues))
+	defer close(inputJobs)
```
Comment (tobiade) [inline chains/chains.go:137]: Nitpick: we can close the `inputJobs` channel immediately after writing inputs to it so we can release the consuming goroutines before the `Apply` method finishes.

File: chains/chains.go [hunk 4]
```diff
@@ -121,6 +121,53 @@ func Predict(ctx context.Context, c Chain, inputValues map[string]any, options .
 	return outputValue, nil
 }
 
+const _defaultApplyMaxNumberWorkers = 5
+
+// Apply executes the chain for each of the inputs asynchronously.
+func Apply(ctx context.Context, c Chain, inputValues []map[string]any, maxWorkers int, options ...ChainCallOption) ([]map[string]any, error) { // nolint:lll
```
Comment (tobiade) [inline chains/chains.go:127]: One for a different PR, but it'd be good to enable timeouts/cancellations using the `ctx` so chains don't run indefinitely (if that ever happens).

Comment (FluffyKebab) [reply chains/chains.go:127]: @tobiade can you check out #178?

Comment (tobiade) [reply chains/chains.go:127]: Thanks @FluffyKebab - left some comments.

**Pull request merged**